### PR TITLE
Add arc length parameterization

### PR DESCRIPTION
**Pull request by Daval-G**
This PR adds a tool to minimize the gradient amplitudes using arc length parameterization. It makes the samples evenly distributed along the path of each shot while maintaining the full shot length, and therefore the maximum gradient amplitude is optimally minimized as all gradient strengths are the same (but not the gradient directions of course). The catch is that the maximum slew rate can become insanely high if the path is not smooth enough. An example is shown below.

Screenshot from 2025-02-15 00-41-06
(A) A classic (blue) spiral trajectory is projected (orange) to be uniformly distributed over the spiral path. (B) shows the gradient amplitudes and (C) shows the slew rates. The projected gradient strength is completely flat but it leads to much stronger slew rates around the center where the spiral direction changes.

### CODE REVIEW

**Comment by Daval-G** [comment]
It appears that the `initialize_3D_annular_shells` function has zero-gradient transitions sometimes, making the interpolation fail. This should not be the case in any trajectory, so for now I am going to make the projection compatible with that case but next MR I will add generic unit tests to ensure none of the trajectories has that

**Comment by paquiteau** [comment]
Nice stuff! I wonder if this can be useful (and maybe supersede) #129 ? Regarding the null gradient problem: From my understanding, it should be equivalent to have the same point sampled consequently multiple times. I guess typically this not desirable, but there might be cases where this could be considered?

**Comment by chaithyagr** [comment]
Awesome stuff. In your short example picture, is it possible to have arc length only above some frequency in kspacr to ensure we don't have slew issues at center due to turning at high speeds

File: src/mrinufft/trajectories/projection.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def fit_arc_length(
+    trajectory: NDArray, order: int | None = None, eps: float = 1e-8
+) -> NDArray:
+    """Adjust the trajectory to have a uniform arc-length distribution.
+
+    The trajectory is parametrized according to its arc length along a
+    cubic-interpolated path and samples are repositioned to minimize
+    the gradients amplitude. This solution is optimal with respect to
+    gradients but can lead to excessive slew rates, and it will change
```
Comment (chaithyagr) [inline src/mrinufft/trajectories/projection.py:17]: Both these points are important and are worth adding as a note or warning

File: src/mrinufft/trajectories/projection.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+"""Functions to fit gradient constraints."""
+
+import numpy as np
+import numpy.linalg as nl
+from numpy.typing import NDArray
+from scipy.interpolate import CubicSpline
+
+
+def fit_arc_length(
```
Comment (chaithyagr) [inline src/mrinufft/trajectories/projection.py:9]: Maybe call it fit_constant_arc_length ?

Comment (chaithyagr) [reply src/mrinufft/trajectories/projection.py:9]: Also it would be worth trying to add support for this parameterization only in certain portions of kspace , but i think it's not easy as you might have discontinuity

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Some first round of reviews... Minor comments for now n some to think about. Thank you for this!

File: src/mrinufft/trajectories/projection.py (hunk 2 above)
Comment (Daval-G) [reply src/mrinufft/trajectories/projection.py:9]: > Maybe call it fit_constant_arc_length ?

Discussing with ChatGPT made me realize using `fit` isn't appropriate, WDYT ? I consider solutions like `resample_by_arc_length`, `make_uniform_arc_length`, `balance_arc_length`, `equalize_arc_length`

Comment (Daval-G) [reply src/mrinufft/trajectories/projection.py:9]: > Also it would be worth trying to add support for this parameterization only in certain portions of kspace , but i think it's not easy as you might have discontinuity

It would be the same as using the function but giving as input just a subpart of the trajectory. Would it be OK with you if I show explicitly how to do that in the example rather than changing the function for something the user can do?

Comment (chaithyagr) [reply src/mrinufft/trajectories/projection.py:9]: I am okay with that surely, but I am worried that we will not be able to handle the discontinuity well.. We potentially will have slew violations there/

File: src/mrinufft/trajectories/trajectory3D.py [hunk 3]
```diff
@@ -728,6 +729,9 @@ def initialize_3D_annular_shells(
     """Initialize 3D trajectories with annular shells.
 
     An exclusive trajectory inspired from the work proposed in [HM11]_.
+    It is similar to other trajectories based on concentric rings but
+    rings are split and recombined by pairs to better balance
+    the shot lengths.
```
Comment (chaithyagr) [inline src/mrinufft/trajectories/trajectory3D.py:734]: I dont fully understand this... split and recombined by pairs? 

Also, can I say that earlier implementation was wrong as this function didn't exist, or did this function make this implementation better?

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Some more comments.. Adding example would be helpful. 
Also @paquiteau its time we start adding examples which is using this function feature at the end of docs. WDYT?

**Comment by paquiteau** [comment]
Let's tackle the doc /example links in another PR (#237 for further discussion)

**Comment by chaithyagr** [comment]
Just noting here that #272 should be able to handle this.

File: src/mrinufft/trajectories/trajectory3D.py (hunk 3 above)
Comment (Daval-G) [reply src/mrinufft/trajectories/trajectory3D.py]: Updated, let me know if it's more clear :)

File: src/mrinufft/trajectories/projection.py (hunk 2 above)
Comment (Daval-G) [reply src/mrinufft/trajectories/projection.py]: After trying it's actually not possible, or more precisely it's far beyond the scope of that method as it would imply working on the slew rate too, which is what #272 is doing more seriously

**Comment by Daval-G** [comment]
@chaithyagr WDYT ?

File: src/mrinufft/trajectories/trajectory3D.py [hunk 4]
```diff
@@ -832,6 +837,8 @@ def initialize_3D_annular_shells(
         trajectory[count : count + Ms] = shell
         count += Ms
 
+    # Balance the samples over the recombined shots
+    # trajectory = parameterize_by_arc_length(trajectory) # FIXME
```
Comment (chaithyagr) [inline src/mrinufft/trajectories/trajectory3D.py:841]: Can you please add a comment explaining the FIXME

**Review by chaithyagr — APPROVED** [review]
Just minor changes. Please update it and if possible open an issue to track it. Otherwise LGTM, just need tests passing. You can merge when green and after handeling this. I dont think we need another review. Thanks again for your wonderful work
